src/target/riscv/riscv-013.c

```diff
@@ -1579,57 +1567,74 @@ static int deassert_reset(struct target *target)
 	RISCV013_INFO(info);
 	select_dmi(target);
 
-	LOG_DEBUG("%d", r->current_hartid);
-
 	/* Clear the reset, but make sure haltreq is still set */
 	uint32_t control = 0;
 	control = set_field(control, DMI_DMCONTROL_HALTREQ, target->reset_halt ? 1 : 0);
-	control = set_field(control, hartsel_mask(target), r->current_hartid);
 	control = set_field(control, DMI_DMCONTROL_DMACTIVE, 1);
-	dmi_write(target, DMI_DMCONTROL, control);
+	dmi_write(target, DMI_DMCONTROL,
+			set_field(control, hartsel_mask(target), r->current_hartid));
 
 	uint32_t dmstatus;
 	int dmi_busy_delay = info->dmi_busy_delay;
 	time_t start = time(NULL);
 
-	if (target->reset_halt) {
-		LOG_DEBUG("Waiting for hart to be halted.");
-		do {
-			if (dmstatus_read(target, &dmstatus, true) != ERROR_OK)
-				return ERROR_FAIL;
-			if (time(NULL) - start > riscv_reset_timeout_sec) {
-				LOG_ERROR("Hart didn't halt coming out of reset in %ds; "
-						"dmstatus=0x%x; "
-						"Increase the timeout with riscv set_reset_timeout_sec.",
-						riscv_reset_timeout_sec, dmstatus);
-				return ERROR_FAIL;
-			}
-			target->state = TARGET_HALTED;
-		} while (get_field(dmstatus, DMI_DMSTATUS_ALLHALTED) == 0);
+	for (int i = 0; i < riscv_count_harts(target); ++i) {
+		int index = i;
+		if (target->rtos) {
+			if (!riscv_hart_enabled(target, index))
+				continue;
+			dmi_write(target, DMI_DMCONTROL,
+					set_field(control, hartsel_mask(target), index));
+		} else {
+			index = r->current_hartid;
+		}
 
-		control = set_field(control, DMI_DMCONTROL_HALTREQ, 0);
-		dmi_write(target, DMI_DMCONTROL, control);
+		if (target->reset_halt) {
+			LOG_DEBUG("Waiting for hart %d to halt out of reset.", index);
+			do {
+				if (dmstatus_read(target, &dmstatus, true) != ERROR_OK)
+					return ERROR_FAIL;
+				if (time(NULL) - start > riscv_reset_timeout_sec) {
+					LOG_ERROR("Hart %d didn't halt coming out of reset in %ds; "
```

Comment:
So, I ran into this on the current master (I think this code is basically the same) when debugging something that takes a few seconds to come out of reset. The `dmstatus_read` following the reset deassertion can actually take the entire time for the `riscv_reset_timeout_sec`.  In other words, the `dmi_read` won't complete until you come out of reset. So then I get this error about "Hart didn't come out of reset", but who cares, because you aren't actually using the user-specified timeout value for that. Anyway in this case I do need to bump the timeout I guess, but it's just sort of misleading. For this particular `dmstatus_read` you should be applying the timeout the user gave you with `set_reset_timeout_sec`. Anyway, shouldn't you be checking `DMI_STATUS_ALLHALTED` before you conclude the hart hasn't come out of reset and return ERROR_FAIL?

Reply:
in other words, something about the order of these checks seems off.

Reply:
I see now that your code above now uses the riscv_command_timeout_sec in the dmi_op loop instead of the hard-coded 256. I like this, someone who is asserting reset can just set riscv_command_timeout_sec = riscv_reseet_timeout_sec temporarily. So I'll just merge this PR into my current one to see if it solves my problem.

Reply:
The general approach seems good, but I think there is still something off in the order of the checks and how you report the fix to the user.